### PR TITLE
feat(config): Make configuration mangement easier. Closes #2463

### LINKED ISSUE #2463: Update configm ap to support strategic-merge-patch
# Summary

Currently the config map has a single field named `config`. This is a PITA to use with Kustomize.

# Motivation

Make it easier to GitOps you config.

# Proposal

Allow you to use normal keys.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the USERS.md.
* [x] I've signed the CLA and required builds are green.

### CODE REVIEW

File: config/controller.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
+		options.FieldSelector = fieldSelector.String()
+		req := restClient.Get().
+			Namespace(cc.namespace).
+			Resource(resource).
+			VersionedParams(&options, metav1.ParameterCodec)
+		return req.Do().Get()
+	}
+	watchFunc := func(options metav1.ListOptions) (watch.Interface, error) {
+		options.Watch = true
```
Comment (alexec) [inline config/controller.go]: lift-and-shift

File: config/controller.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+			VersionedParams(&options, metav1.ParameterCodec)
+		return req.Watch()
+	}
+	source := &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+	_, controller := cache.NewInformer(
+		source,
+		&apiv1.ConfigMap{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(old, new interface{}) {
```
Comment (alexec) [inline config/controller.go]: removed add func here - need to check this is ok

**Comment by alexec** [comment]
@jessesuen @sarabala1979 could I get your feedback on this concept? I've always wanted to use kustomize to be able to merge Argo configuration. This basically allows you to do that. It's most obviously show in the new example config map:

https://github.com/argoproj/argo/blob/54d47470943de25ade01d9ca6ea9224d48f30662/docs/workflow-controller-configmap.yaml

File: config/controller.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	return onChange(c)
+}
+
+func parseConfigMap(cm *apiv1.ConfigMap) (Config, error) {
+	// The key in the configmap to retrieve workflow configuration from.
+	// Content encoding is expected to be YAML.
+	var c Config
+	config, ok := cm.Data["config"]
+	if ok && len(cm.Data) != 1 {
+		return c, fmt.Errorf("if you have an item in your config map named 'config', you must only have one item")
```
Comment (alexec) [inline config/controller.go]: should we allow mix-and-match?

**Comment by alexec** [comment]
@simster7 and @sarabala1979 ready for review

**Comment by simster7** [comment]
This is backwards compatible right?

File: workflow/controller/config_controller.go [hunk 4]
```diff
@@ -2,49 +2,21 @@ package controller
...
-	if err != nil {
-		return errors.InternalWrapError(err)
-	}
-	log.Printf("workflow controller configuration from %s:\n%s", wfc.configMap, configStr)
+func (wfc *WorkflowController) updateConfig(config config.Config) error {
 	if wfc.cliExecutorImage == "" && config.ExecutorImage == "" {
-		return errors.Errorf(errors.CodeBadRequest, "ConfigMap '%s' does not have executorImage", wfc.configMap)
+		return errors.Errorf(errors.CodeBadRequest, "ConfigMap does not have executorImage")
 	}
 	wfc.Config = config
```
Comment (alexec) [inline workflow/controller/config_controller.go]: should log new config here at INFO

**Pull request merged**